### PR TITLE
FO password reminder test with maildev

### PR DESCRIPTION
**Pull request by SD1982**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add a new test "password reminder"
| Type?             | refacto 
| Category?         | TE 
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | //.
| How to test?      | Can be tested with this command `TEST_PATH="functional/FO/01_login/01_passwordReminder" URL_FO="your FO url" npm run specific-test
| Possible impacts? | Nothing only tests.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22932)

### CODE REVIEW

File: tests/UI/pages/FO/passwordReminder/index.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  /*
+  Methods
+   */
+
+  /** Fill the reset password email form field and click "send reset link" button
+   *
+   * @param page
+   * @param email
+   * @returns {Promise<void>}
```
Comment (boubkerbribri) [inline tests/UI/pages/FO/passwordReminder/index.js]:
```suggestion
  /** 
   * Fill the reset password email form field and click "send reset link" button
   * @param page
   * @param email
   * @returns {Promise<void>}
```

File: tests/UI/pages/FO/passwordReminder/index.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+   */
+  async sendResetPasswordLink(page, email) {
+    await this.setValue(page, this.emailFormField, email);
+    await this.clickAndWaitForNavigation(page, this.sendResetLinkButton);
+  }
+
+  /** Check that the success alert message is visible
+   *
+   * @param page
+   * @returns {Promise<boolean>}
```
Comment (boubkerbribri) [inline tests/UI/pages/FO/passwordReminder/index.js]:
```suggestion
  /** 
   * Check that the success alert message is visible
   * @param page
   * @returns {Promise<boolean>}
```

File: themes/classic/templates/customer/password-email.tpl [hunk 3]
```diff
@@ -54,7 +54,7 @@
         <div class="col-md-5 email">
           <input type="email" name="email" id="email" value="{if isset($smarty.post.email)}{$smarty.post.email|stripslashes}{/if}" class="form-control" required>
         </div>
-        <button class="form-control-submit btn btn-primary hidden-xs-down" name="submit" type="submit">
+        <button id="send_reset_link" class="form-control-submit btn btn-primary hidden-xs-down" name="submit" type="submit">
```
Comment (boubkerbribri) [inline themes/classic/templates/customer/password-email.tpl]:
```suggestion
        <button id="send-reset-link" class="form-control-submit btn btn-primary hidden-xs-down" name="submit" type="submit">
```

File: themes/classic/templates/customer/password-email.tpl [hunk 4]
```diff
@@ -67,7 +67,7 @@
 {/block}
 
 {block name='page_footer'}
-  <a href="{$urls.pages.my_account}" class="account-link">
+  <a id="back_to_login" href="{$urls.pages.my_account}" class="account-link">
```
Comment (boubkerbribri) [inline themes/classic/templates/customer/password-email.tpl]:
```suggestion
  <a id="back-to-login" href="{$urls.pages.my_account}" class="account-link">
```

File: tests/UI/pages/FO/passwordReminder/index.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+class PasswordReminder extends FOBasePage {
+  constructor() {
+    super();
+
+    this.pageTitle = 'Forgot your password';
+
+    // Selectors
+    this.emailFormField = '#email';
+    this.backToLoginLink = '#back_to_login'
+    this.sendResetLinkButton = '#send_reset_link';
```
Comment (boubkerbribri) [inline tests/UI/pages/FO/passwordReminder/index.js]:
```suggestion
    this.backToLoginLink = '#back-to-login'
    this.sendResetLinkButton = '#send-reset-link';
```

File: tests/UI/pages/BO/advancedParameters/email/index.js [hunk 6]
```diff
@@ -194,6 +203,39 @@ class Email extends BOBasePage {
...
+
+  /**
+   * Reset the mail parameters to default config
+   * @param page
+   * @returns {Promise<string>}
+   */
+  async resetDefaultParameters(page) {
+    // Click on smtp radio button
+    await page.click(this.sendMailParametersRadioButton);
+    await this.elementNotVisible(page, this.smtpServerFormField);
```
Comment (boubkerbribri) [inline tests/UI/pages/BO/advancedParameters/email/index.js]:
```suggestion
    await page.waitForSelector(this.smtpServerFormField, {state: hidden});
```
The function `elementNotVisible` does not throw exception if the element is not visible.

File: tests/UI/campaigns/commonTests/loginFO.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+const myAccountPage = require('@pages/FO/myAccount');
+
+module.exports = {
+  async loginFO(mochaContext, page, customer) {
+    await testContext.addContextItem(mochaContext, 'testIdentifier', 'loginFO');
+    await loginPage.goTo(page, global.FO.URL);
+    await myAccountPage.goToLoginPage(page);
+    await loginPage.customerLogin(page, customer);
+    const pageTitle = await myAccountPage.getPageTitle(page);
+    await expect(pageTitle).to.contains(myAccountPage.pageTitle);
```
Comment (boubkerbribri) [inline tests/UI/campaigns/commonTests/loginFO.js:16]: Check that the customer is connected before checking page title

File: tests/UI/campaigns/commonTests/loginFO.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  },
+
+  async logoutFO(mochaContext, page) {
+    await testContext.addContextItem(mochaContext, 'testIdentifier', 'logoutFO');
+    await loginPage.goToMyAccountPage(page);
+    await loginPage.logout(page);
+    const pageTitle = await loginPage.getPageTitle(page);
+    await expect(pageTitle).to.contains(loginPage.pageTitle);
```
Comment (boubkerbribri) [inline tests/UI/campaigns/commonTests/loginFO.js]: Check that the customer is not connected before checking page title

File: tests/UI/campaigns/commonTests/loginFO.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+require('module-alias/register');
+const {expect} = require('chai');
+const testContext = require('@utils/testContext');
+const loginPage = require('@pages/FO/login');
+const myAccountPage = require('@pages/FO/myAccount');
+
+module.exports = {
+  async loginFO(mochaContext, page, customer) {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/commonTests/loginFO.js:8]: If We add common tests about FO, we should update one checked identifiers `tools/stepIdentifiersChecker.js`.

**Review by boubkerbribri — COMMENTED** [review]
Comments about test

File: tests/UI/README.md [hunk 10]
```diff
@@ -35,7 +35,12 @@ npm install
 | DB_USER             | Login user of your MySql (default to **`root`**) |
 | DB_NAME             | Name of the MySql database (default to **`prestashop_db`**) |
 | DB_PASSWD           | Password for your MySql (default to **`empty`**) |
+### Maildev parameters
 
+| Parameter           | Description                                          |
+|---------------------|----------------------------------------------------- |
+| SMTP_SERVER             | The smtp server address for maildev (default to **`172.20.0.4`**)|
```
Comment (SimonGrn) [inline tests/UI/README.md:42]: This is a weird default value. Any reason ?

File: tests/UI/campaigns/tools/stepIdentifiersChecker.js [hunk 11]
```diff
@@ -23,7 +23,7 @@ const count = contexts => contexts.reduce(
  */
 const duplicates = dict => Object.keys(dict)
   .filter(
-    key => dict[key] > 1 && key !== 'loginBO' && key !== 'logoutBO',
+    key => dict[key] > 1 && key !== 'loginBO' && key !== 'logoutBO' && key !== 'logoutFO' && key !== 'logoutFO',
```
Comment (SimonGrn) [inline tests/UI/campaigns/tools/stepIdentifiersChecker.js]:
```suggestion
    key => dict[key] > 1 && key !== 'loginBO' && key !== 'logoutBO' && key !== 'loginFO' && key !== 'logoutFO',
```

File: tests/UI/README.md (hunk 10 above)
Comment (SD1982) [reply tests/UI/README.md:42]: it's to communicate with the container localhost when the test run on the docker

File: tests/UI/campaigns/functional/FO/01_login/01_passwordReminder.js [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+      const pageTitle = await emailPage.getPageTitle(page);
+      await expect(pageTitle).to.contains(emailPage.pageTitle);
+    });
+
+    it('should fill the smtp parameters form fields', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'fillSmtpParametersFormField', baseContext);
+
+      const alertSuccessMessage = await emailPage.setupSmtpParameters(page, smtpServer, DefaultAccount.email,
+        DefaultAccount.password, smtpPort);
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/FO/01_login/01_passwordReminder.js]:
```suggestion
      const alertSuccessMessage = await emailPage.setupSmtpParameters(
        page, 
        smtpServer, 
        DefaultAccount.email,
        DefaultAccount.password, 
        smtpPort,
      );

```

File: tests/UI/campaigns/utils/mailHelper.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
+const MailDev = require('maildev');
+
+module.exports = {
+  /**
+   *
```
Comment (boubkerbribri) [inline tests/UI/campaigns/utils/mailHelper.js:7]: Please add description for the function

File: tests/UI/campaigns/utils/mailHelper.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+   * @param config
+   * @returns {*}
+   */
+  createMailListener(config = global.maildevConfig) {
+    return new MailDev({
+      smtp: config.smtpPort,
+    });
+  },
+  /**
+   *
```
Comment (boubkerbribri) [inline tests/UI/campaigns/utils/mailHelper.js:18]: same here

File: tests/UI/campaigns/utils/mailHelper.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+   */
+  startListener(mailListener) {
+    mailListener.listen((err) => {
+      if (err) {
+        throw new Error(err);
+      }
+    });
+  },
+  /**
+   *
```
Comment (boubkerbribri) [inline tests/UI/campaigns/utils/mailHelper.js:30]: same here

File: tests/UI/campaigns/functional/FO/01_login/01_passwordReminder.js (hunk 12 above)
Comment (SD1982) [reply tests/UI/campaigns/functional/FO/01_login/01_passwordReminder.js]: done !

File: tests/UI/campaigns/functional/FO/01_login/01_passwordReminder.js [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+// Import test context
+const testContext = require('@utils/testContext');
+
+const baseContext = 'functional_FO_login_passwordReminder';
+
+let browserContext;
+let page;
+let newMail;
+const {smtpServer} = global.maildevConfig;
+const {smtpPort} = global.maildevConfig;
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/FO/01_login/01_passwordReminder.js]:
```suggestion
const {smtpServer, smtpPort} = global.maildevConfig;
```

File: tests/UI/campaigns/tools/stepIdentifiersChecker.js [hunk 17]
```diff
@@ -23,7 +23,7 @@ const count = contexts => contexts.reduce(
  */
 const duplicates = dict => Object.keys(dict)
   .filter(
-    key => dict[key] > 1 && key !== 'loginBO' && key !== 'logoutBO',
+    key => dict[key] > 1 && key !== 'loginBO' && key !== 'logoutBO' && key !== 'loginFO' && key !== 'logoutFO',
```
Comment (Progi1984) [inline tests/UI/campaigns/tools/stepIdentifiersChecker.js]:
```suggestion
    key => dict[key] > 1 && ['loginBO', 'logoutBO', 'loginFO', 'logoutFO'].indexOf(key) === -1,
```

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks

**Review by Progi1984 — APPROVED** [review]

**Review by boubkerbribri — APPROVED** [review]

**Comment by Progi1984** [comment]
Tahnks @SD1982

**Pull request merged**